File: app/src/main/java/rikka/akashitoolkit/event/Event.java
package rikka.akashitoolkit.event;

import android.support.annotation.Nullable;
import android.util.Log;

import com.google.gson.Gson;

import java.util.ArrayList;
import java.util.List;

import rikka.akashitoolkit.model.LocaleMultiLanguageEntry;
import rikka.akashitoolkit.model.MultiLanguageEntry;
import rikka.akashitoolkit.model.ShipVoice;
import rikka.akashitoolkit.utils.FlavorsUtils;
import rikka.akashitoolkit.utils.Utils;

/**
 * Created by Rikka on 2016/8/12.
 */
public class Event extends ArrayList<Event.Container> {

    public static final int API_VERSION = 1;

    public static final int TYPE_TITLE = 0;
    public static final int TYPE_GALLERY = 1;
    public static final int TYPE_CONTENT = 2;
    public static final int TYPE_URL = 3;
    public static final int TYPE_VOICE = 4;

    public static class Container {

        private int type;
        private Object object; // LinkedTreeMap

        private boolean only_play;
        private boolean only_china;
        private boolean not_safe; // Google: 不可以，这很色情

        private int min_api;
        private int max_api;

        public int getType() {
            return type;
        }

        public Object getObject() {
            return object;
        }

        /**
         * 根据 API 版本，是否是 Play 版 / 国内版决定是否显示
         *
         * @return 是否显示
         */
        public boolean shouldShow() {
            return !((max_api != 0 && API_VERSION > max_api) || (min_api != 0 && API_VERSION < min_api)) && !(only_china && FlavorsUtils.isPlay()) && !(only_play && !FlavorsUtils.isPlay()) && !(not_safe && FlavorsUtils.shouldSafeCheck());
        }
    }

    public interface OnItemParseListener {
        void onItemParse(int type, Object object);
    }

    public void parse(@Nullable OnItemParseListener listener) {
        Gson gson = new Gson();
        for (Container container : this) {
            Object object = null;
            String json = gson.toJson(container.getObject());

            if (!container.shouldShow()) {
                //Log.d("Event", "skipped: " + json);
                continue;
            }

            boolean error = false;
            switch (container.getType()) {
                case TYPE_TITLE:
                case TYPE_CONTENT:
                    object = gson.fromJson(json, Event.Title.class);
                    break;
                case TYPE_GALLERY:
                    object = gson.fromJson(json, Event.Gallery.class);

                    Event.Gallery _data = (Event.Gallery) object;

                    List<String> urls = new ArrayList<>();
                    for (String url : _data.getUrls()) {
                        if (!url.startsWith("http")) {
                            urls.add(Utils.getKCWikiFileUrl(url));
                        } else {
                            urls.add(url);
                        }
                    }
                    _data.getUrls().clear();
                    _data.getUrls().addAll(urls);
                    break;
                case TYPE_URL:
                    object = gson.fromJson(json, Event.Url.class);
                    break;
                default:
                    Log.e("Event", "unhandled: " + json);
                    error = true;
            }

            if (!error && listener != null) {
                listener.onItemParse(container.getType(), object);
            }
        }
    }

    public static class BaseObject {
        private LocaleMultiLanguageEntry title;
        private LocaleMultiLanguageEntry summary;
        private LocaleMultiLanguageEntry content;

        public MultiLanguageEntry getTitle() {
            if (title == null) {
                title = new LocaleMultiLanguageEntry();
            }

            return title;
        }

        public LocaleMultiLanguageEntry getSummary() {
            if (summary == null) {
                summary = new LocaleMultiLanguageEntry();
            }

            return summary;
        }

        public MultiLanguageEntry getContent() {
            if (content == null) {
                content = new LocaleMultiLanguageEntry();
            }

            return content;
        }
    }

    /**
     * 标题
     */
    public static class Title extends BaseObject {

    }

    /**
     * 具有一个按钮可以打开一个链接
     */
    public static class Url extends BaseObject {
        private String url;
        private LocaleMultiLanguageEntry url_text;

        public String getUrl() {
            return url;
        }

        public MultiLanguageEntry getUrlText() {
            return url_text;
        }
    }

    /**
     * 图片们
     */
    public static class Gallery extends BaseObject {
        List<String> urls;
        // TODO support multi language ?
        List<String> names;
        List<Integer> ids;
        int action_type;

        public List<String> getUrls() {
            return urls;
        }

        public List<String> getNames() {
            return names;
        }

        public List<Integer> getIds() {
            return ids;
        }

        public int getActionType() {
            return action_type;
        }
    }

    /**
     * 语音们
     */
    public static class Voices extends BaseObject {
        private List<Voice> voices;

        public static class Voice extends BaseObject {
            private LocaleMultiLanguageEntry type;
            private List<ShipVoice> voice;

            public LocaleMultiLanguageEntry getType() {
                return type;
            }

            public List<ShipVoice> getVoice() {
                return voice;
            }
        }

        public List<Voice> getVoices() {
            return voices;
        }
    }
}

File: server/event.php
<?php
$api_version = $_GET['api_version'];

$TYPE_TITLE = 0;
$TYPE_GALLERY = 1;
$TYPE_CONTENT = 2;
$TYPE_URL = 3;
$TYPE_VOICE = 4;

$json = array();

$data = null;
$data['type'] = $TYPE_TITLE;
$data['object']['title']['zh_cn'] = "2018年冬季活动";
$data['object']['content']['zh_cn'] = "捷号決戦！邀撃、レイテ沖海戦(後篇)";
array_push($json, $data);

$data = null;
$data['type'] = $TYPE_URL;
$data['object']['title']['zh_cn'] = "详细攻略";
$data['object']['content']['zh_cn'] = "";
$data['object']['url_text']['zh_cn'] = "kcwiki:2018年冬季活动";
$data['object']['url'] = "https://zh.kcwiki.org/wiki/2018%E5%B9%B4%E5%86%AC%E5%AD%A3%E6%B4%BB%E5%8A%A8";
array_push($json, $data);

$data = null;
$data['type'] = $TYPE_TITLE;
$data['object']['title']['zh_cn'] = "活动原型";
$data['object']['content']['zh_cn'] = "莱特湾海战";
array_push($json, $data);

$data = null;
$data['type'] = $TYPE_CONTENT;
$data['object']['title']['zh_cn'] = "概述";
$data['object']['content']['zh_cn'] = "莱特湾海战发生于1944年10月20日，到10月26日结束，是当时发生在菲律宾周围日本联合舰队与美国海军和澳大利亚海军组成的联合国军之间一系列的海战的总称。这些海战从前到后可以大概分为：巴拉望水道海战、锡布延海战、苏里高海峡海战、恩加尼奥角海战、萨马岛海战。

上下6天之内，在这场美军称作KING II行动，日方称作捷一号作战的海战中。日本联合舰队投入了尚能执行任务的舰艇中的多数；美国海军也投入了其在太平洋上活动的大半军舰。双方总计投入21艘航空母舰、21艘战列舰、170艘驱逐舰与近2,000架飞机。因此，莱特湾海战凭借其规模之大、参战舰艇之多、作战范围之广，获得了史上最大规模海战的称呼。
";
array_push($json, $data);

$data = null;
$data['type'] = $TYPE_URL;
$data['object']['title']['zh_cn'] = "掉落统计";
$data['object']['content']['zh_cn'] = "POI2018年冬季活动统计";
$data['object']['url'] = "https://db.kcwiki.org/drop/";
$data['object']['url_text']['zh_cn'] = "poi-statistics掉落统计";
array_push($json, $data);

echo json_encode($json);